tests/run/cpp_iterators_simple.h

```diff
@@ -8,3 +8,14 @@ class DoublePointerIter {
     int len_;
 };
 
+class DoublePointerIterCanMakeEmpty {
+    // an alternate version that is default-constructable
+public:
+    DoublePointerIterCanMakeEmpty() {}
+    DoublePointerIterCanMakeEmpty(double* start, int len) : start_(start), len_(len) { }
+    double* begin() { return start_; }
+    double* end() { return start_ + len_; }
+private:
+    double* start_;
+    int len_;
+};
```

Comment:
Why not inherit from `DoublePointerIter`?
```suggestion
class DoublePointerIterCanMakeEmpty: public DoublePointerIter {
    // an alternate version that is default-constructable
public:
    DoublePointerIterCanMakeEmpty() {}
};
```

Reply:
Also, why is this class needed at all? And would it be clerer if it were named something like
* `DoublePointerIterCanBeEmpty`
* `DoublePointerIterDefaultConstructible`
* ...
?

Reply:
If I remember correctly, the class is needed because an instance is default-constructed by value on the internal closure class.

(There's a separate debate here about whether to capture C++ references by value or reference in the closure of course, but I've gone for value as the safe option).

Either way, with this choice the captured iterator does need to be default constructable to compile.

Your change is a good idea but it hides the base class constructor. I'll propose a different change that should fix it

```suggestion
class DoublePointerIterDefaultConstructible: public DoublePointerIter {
    // an alternate version that is default-constructible
public:
    DoublePointerIterDefaultConstructible() {}
    DoublePointerIterDefaultConstructible(double* start, int len) :
        DoublePointerIter(start, len)
    {}
    
};
```